### PR TITLE
fix: create verifiable Standard JSON for projects with external files

### PR DESCRIPTION
**Pull request by tash-2s**
- Resolves https://github.com/foundry-rs/foundry/issues/5307

Currently, Foundry projects containing Solidity files outside the project root directory face contract verification failures on block explorers. This issue occurs from the Standard JSON including unusable source paths for external files, represented as full absolute paths in their host file systems.

This PR addresses the issue by improving the path conversion process. For files not located under the project root directory, relative parent directory paths (`..`) are used, enabling the compiler to find the files within the json.

Steps to reproduce the issue are detailed in the linked issue above. Additionally, a test case representing that scenario has been added.

With this change, the json created in the reproduction scenario will appear as follows:

```json
{
  "language": "Solidity",
  "sources": {
    "src/Counter.sol": {
      "content": "// SPDX-License-Identifier: UNLICENSED\npragma solidity ^0.8.13;\n\nimport \"@remapped/Parent.sol\";\n\ncontract Counter {\n    uint256 public number;\n\n    function setNumber(uint256 newNumber) public {\n        number = newNumber;\n    }\n\n    function increment() public {\n        number++;\n    }\n}\n"
    },
    "../remapped/Parent.sol": {
      "content": "// SPDX-License-Identifier: UNLICENSED\npragma solidity ^0.8.13;\nimport \"./Child.sol\";\n"
    },
    "../remapped/Child.sol": {
      "content": "// SPDX-License-Identifier: UNLICENSED\npragma solidity ^0.8.13;\n"
    }
  },
  "settings": {
    "remappings": [
      "@remapped/=../remapped/",
      "ds-test/=lib/forge-std/lib/ds-test/src/",
      "forge-std/=lib/forge-std/src/"
    ],
    "optimizer": {
      "enabled": true,
      "runs": 200
    },
    "metadata": {
      "useLiteralContent": false,
      "bytecodeHash": "ipfs",
      "appendCBOR": true
    },
    "outputSelection": {
      "*": {
        "": [
          "ast"
        ],
        "*": [
          "abi",
          "evm.bytecode",
          "evm.deployedBytecode",
          "evm.methodIdentifiers",
          "metadata"
        ]
      }
    },
    "evmVersion": "paris",
    "libraries": {}
  }
}
```

The source path is now aligned with the project root.

I have successfully deployed and verified the contract on Etherscan using this change.

`forge create --rpc-url "wss://ethereum-holesky.publicnode.com" --verify --verifier-url "https://api-holesky.etherscan.io/api" --etherscan-api-key "..." --private-key "..." src/Counter.sol:Counter`

https://holesky.etherscan.io/address/0xe08c332706185521fc8bc2b224f67adf814b1880#code

### CODE REVIEW

File: tests/project.rs
Comment (tash-2s) [inline tests/project.rs:1617]: After [my previous PR](https://github.com/foundry-rs/compilers/pull/35) is merged, this canonicalization can be removed. It is necessary on this branch because the symlink-resolving canonicalization is still in effect, and the temporary directory on macOS is a symlink.

**Review by mattsse — COMMENTED** [review]
amazing, tysm!

as mentioned in the other PR, I'd like to see a foundry companion PR with patched crate dep as a sanity check before we merge this

**Comment by tash-2s** [comment]
Thank you for your review!

I've created [a companion PR in foundry](https://github.com/foundry-rs/foundry/pull/6676) with this patch. Also, I have confirmed that the tests in foundry passed successfully.

**Comment by tash-2s** [comment]
The workflow test for this PR failed due to a failure in nextest installation. As this issue is unrelated to the changes made, re-running the test should resolve it. https://github.com/foundry-rs/compilers/actions/runs/7331166894/job/19979490471

File: src/lib.rs
Comment (mattsse) [inline src/lib.rs:980]: I think we can solve this a bit differently
don't zip the two iterators but create the two iters separately and then iterate over the `path`
and continue as long as base and path components match

this way we don't need the vecdeque

**Review by mattsse — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (tash-2s) [reply src/lib.rs:980]: Hmm, I'm not sure I understand your suggestion.

If we compare the components of base and path at the same positions, it seems similar to using zip.

Could you please elaborate on the type of implementation you have in mind?

Comment (mattsse) [reply src/lib.rs:980]: something like

```rust
let mut components = Vec::new(); // perhaps we can even directly push components to a PathBuf?
let mut base_iter = base.components();
let mut path_iter = path.components();

while let Some(path) = path_iter.next() {
  if Some(path) != path_iter.next() {
     components.push(path);
     components.extend(path_iter);
  }
}


```

Comment (tash-2s) [reply src/lib.rs]: Thank you for your explanation. I now understand your point. I've made the change and pushed it.

File: src/lib.rs
Comment (mattsse) [inline src/lib.rs:958]: can we add an example here that illustrates this, otherwise this is a bit hard to understand.

**Review by mattsse — CHANGES_REQUESTED** [review]
last pedantic doc nit

File: src/lib.rs
Comment (tash-2s) [reply src/lib.rs]: I've updated the comment, adding examples to make it easier to understand.

**Review by mattsse — APPROVED** [review]
awesome job!

I'll merge both and upstream to foundry after a new release

**Comment by PaulRBerg** [comment]
Amazing work.

Doesn't this PR also address issue https://github.com/foundry-rs/foundry/issues/3507?

**Pull request merged**
